rmw_cyclonedds_cpp/src/rmw_node.cpp

```diff
@@ -510,39 +620,111 @@ extern "C" rmw_ret_t rmw_guard_condition_set_listener_callback(
+
+  // This API sets two callbacks:
+  // 1. The DDS guard condition's listener callback.
+  // 2. The user callback, called from the listener callback.
+  /*
+      dds_listener_callback(..) {
+        user_callback(..);
+      }
+  */
+
+  // Set the user callback data
+  user_callback_data_t * data = &(gc->user_callback_data);
+
+  if (user_data && guard_condition_handle && callback) {
+    data->callback = callback;
+    data->user_data = user_data;
+    data->event_type = WAITABLE_EVENT;
+    data->entity_handle = guard_condition_handle;
+  } else {
+    // Unset callback: If any of the user callback pointers is NULL,
+    // assign a NULL listener to the entity.
+    return dds_set_listener(entity_to_listen, NULL);
+  }
+
+  // Create a listener which will use the user data
+  // as argument when calling its callback
+  gc->listener = dds_create_listener(data);
+
+  // Assign the DDS listener callback to the
+  // guard condition listener
+  dds_lset_data_on_readers(gc->listener, dds_listener_callback);
+
+  // Finally attach the listener to the guard condition
+  // NOTE: The listener doesn't seem to listen when the guard condition is triggered!
+  // so there's a hack on rmw_trigger_guard_condition() to call the callback
+  // manually
+  return dds_set_listener(entity_to_listen, gc->listener);
 }
 
-extern "C" rmw_ret_t rmw_event_set_events_listener_callback(
-  const void * user_data,
+extern "C" rmw_ret_t rmw_event_set_listener_callback(
+  rmw_event_t * rmw_event,
   rmw_listener_callback_t callback,
+  const void * user_data,
   const void * waitable_handle,
-  rmw_event_t * rmw_event,
   bool use_previous_events)
 {
-  (void)user_data;
-  (void)callback;
-  (void)waitable_handle;
-  (void)rmw_event;
-  (void)use_previous_events;
-  // auto event = static_cast<CddsEvent *>(rmw_event->data);
-  // event->setCallback(user_data, callback,
-  //                              waitable_handle, use_previous_events);
-  RCUTILS_LOG_ERROR_NAMED(
-    "rmw_node.cpp",
-    "rmw_event_set_events_listener_callback: not supported (yet)");
-  return RMW_RET_UNSUPPORTED;
+  auto dds_event = static_cast<CddsEvent *>(rmw_event->data);
+  auto entity_to_listen = dds_event->enth;
+
+  // This API sets two callbacks:
+  // 1. The DDS event's listener callback.
+  // 2. The user callback, called from the listener callback.
+  /*
+      dds_listener_callback(..) {
+        user_callback(..);
+      }
+  */
+
+  // Seems there is a dds_event for each dds_subscription,
+  // and they share the same `enth`, so:
+  //     dds_event->enth = dds_subscription->enth
+  // So setting callbacks to this event overrides the
+  // callbacks already set for the subscription.
+  // For now, we'll check if the event doesn't have an
+  // associated listener we'll assign one here.
+
+  // First, lets check if we don't have already a listener.
+
+  // Note: using dds_get_listener(..) when there's no listener
+  // results in a segfault.
```

Comment:
this is strange, see the code of the function:
```
dds_return_t dds_get_listener (dds_entity_t entity, dds_listener_t *listener)
{
  dds_entity *e;
  dds_return_t ret;
  if (listener == NULL)
    return DDS_RETCODE_BAD_PARAMETER;
  else if ((ret = dds_entity_pin (entity, &e)) != DDS_RETCODE_OK)
    return ret;
  else
  {
    ddsrt_mutex_lock (&e->m_observers_lock);
    dds_copy_listener (listener, &e->m_listener);
    ddsrt_mutex_unlock (&e->m_observers_lock);
    dds_entity_unpin (e);
    return DDS_RETCODE_OK;
  }
}
```

Reply:
The issue was on `dds_copy_listener (listener, &e->m_listener);` 
Anyway this was just a way to not assign listeners to events, because the override the subscriptions listeners.
The idea is ask how to have different entities for events and subscriptions, then the check for listeners already assigned should dissapear

Reply:
Just as you say in other comment, `m_listener` is the listener of the entity. Which when you didn't assign a listener before, is a pointer pointing to nowhere, which could cause the issue in the operation `dds_copy_listener` when dereferencing the not valid ptr:
```
void dds_copy_listener (dds_listener_t * __restrict dst, const dds_listener_t * __restrict src)
{
  if (dst && src)
    *dst = *src;
}
```